sbi/samplers/vi/vi_pyro_flows.py

```diff
@@ -0,0 +1,635 @@
+    return [split_dim, arn], {"log_scale_min_clip": -3.0}
+
+
+def init_spline_coupling(dim, device: str = "cpu", **kwargs):
+    """Intitialize a spline coupling transform, by providing necessary args and
+    kwargs."""
+    assert dim > 1, "In 1d this would be equivalent to affine flows, use them!"
+    split_dim = kwargs.get("split_dim", dim // 2)
+    hidden_dims = kwargs.pop("hidden_dims", [5 * dim + 5, 5 * dim + 5])
+    nonlinearity = kwargs.pop("nonlinearity", nn.ReLU())
+    count_bins = kwargs.get("count_bins", 10)
+    order = kwargs.get("order", "linear")
+    bound = kwargs.get("bound", 5)
+    if order == "linear":
+        param_dims = [
+            (dim - split_dim) * count_bins,
+            (dim - split_dim) * count_bins,
+            (dim - split_dim) * (count_bins - 1),
+            (dim - split_dim) * count_bins,
+        ]
+    else:
+        param_dims = [
+            (dim - split_dim) * count_bins,
+            (dim - split_dim) * count_bins,
+            (dim - split_dim) * (count_bins - 1),
+        ]
+    neural_net = DenseNN(
+        split_dim, hidden_dims, param_dims, nonlinearity=nonlinearity
+    ).to(device)
+    return [dim, split_dim, neural_net], {
+        "count_bins": count_bins,
+        "bound": bound,
+        "order": order,
+    }
+
+
+# Register these directly from pyro
+
+register_transform(
+    transforms.AffineAutoregressive,
+    "affine_autoregressive",
+    inits=init_affine_autoregressive,
+)
+register_transform(
+    transforms.SplineAutoregressive,
+    "spline_autoregressive",
+    inits=init_spline_autoregressive,
+)
+
+register_transform(
+    transforms.AffineCoupling, "affine_coupling", inits=init_affine_coupling
+)
+
+register_transform(
+    transforms.SplineCoupling, "spline_coupling", inits=init_spline_coupling
+)
+
+
+# Register these very simple transforms.
+
+
+@register_transform(
+    name="affine_diag",
+    inits=lambda dim, device="cpu", **kwargs: (
+        [],
+        {
+            "loc": torch.zeros(dim, device=device),
+            "scale": torch.ones(dim, device=device),
+        },
+    ),
+)
+class AffineTransform(transforms.AffineTransform):
+    """Trainable version of an Affine transform. This can be used to get diagonal
+    Gaussian approximations."""
+
+    __doc__ += transforms.AffineTransform.__doc__
+
+    def parameters(self):
+        self.loc.requires_grad_(True)
+        self.scale.requires_grad_(True)
+        yield self.loc
+        yield self.scale
+
+    def with_cache(self, cache_size=1):
+        if self._cache_size == cache_size:
+            return self
+        return AffineTransform(self.loc, self.scale, cache_size=cache_size)
+
+
+@register_transform(
+    name="affine_tril",
+    inits=lambda dim, device="cpu", **kwargs: (
+        [],
+        {
+            "loc": torch.zeros(dim, device=device),
+            "scale_tril": torch.eye(dim, device=device),
+        },
+    ),
+)
+class LowerCholeskyAffine(transforms.LowerCholeskyAffine):
```

Comment:
Why do we need this?

Reply:
This is for full Gaussian approximation. We can also remove it.

Reply:
Let's keep it. Might be nice to have